Environment/depfile: fix bug with Git hash versions

scheibelp opened the pull request:
If a spec contains a `hash=number` version, this was resulting in malformed Makefiles generated by `spack env depfile` (the target cannot include an un-escaped '=' character); for now just remove the version from these targets altogether

scheibelp committed:
if a spec contains a hash=number version, this was resulting in malformed Makefiles generated by 'spack env depfile' (the target cannot include an un-escaped '=' character); for now just remove the version from these targets altogether

scheibelp commented:
> Alternatively, can you map all forbidden characters to _?

Yes: although IMO that is of limited use: the resulting string would be something like `packagename-longhash_number-spackhash` which I consider hard to read (i.e. 72 characters of hashes). I figured since you already went through the effort of adding comments with the full spec, that this part of the identifier was unnecessary (the version is useful in spaces where the same package is being built multiple times, but I'm not sure a reader of the autogenerated makefile would want to be checking in that level of detail).

Do you think a suitable compromise would be to augment those comments with a clear search-able identifier (like "zlib:first7-characters-of-spec-hash") at the beginning of the comment which includes the formatted spec string? Or maybe this can just skip formatting the version for git-hash-based versions?

haampie commented:
> the version is useful in spaces where the same package is being built multiple times, but I'm not sure a reader of the autogenerated makefile would want to be checking in that level of detail).

It's user facing in tab completion, so putting a version there is helpful:

```
$ spack -e . env depfile -o Makefile --make-prefix x
$ make x/install/p<tab>
x/install/perl-5.36.0-t2y654th7mwtzwforeiiqcl7qgvmnazx
x/install/pkgconf-1.8.0-glgyj6nugckh4btiyi5bcwnisiofhgmn
```

scheibelp committed:
trying to formalize the logic by which specs are converted to targets; unfortunately the .adjacency list is directly rendered to template

scheibelp committed:
adjacency list needs to be preformatted targets since they are rendered directly into the template: headache

scheibelp committed:
full test

scheibelp committed:
auto style fix plus flake complaints about unused vars and forgetting to create distinct name for test

scheibelp committed:
allow dashes in spec portion of makefile target; improve docstring in format function to account for lack of clarity on the allowed syntax of makefile targets

scheibelp committed:
flake fix

scheibelp commented:
> Alternatively, can you map all forbidden characters to _?

To be precise, the notion of "forbidden" characters is more appropriate for makefile variable names, which have a whitelist policy (see [here](https://www.gnu.org/software/autoconf/manual/autoconf-2.63/html_node/Special-Chars-in-Names.html)).

The main problem this is trying to solve is `Spec.format` introducing characters which produce technically valid but meaningless makefiles. To that end, the goal would be to strip any character which has a special meaning in `make`.

To be safe, I updated the logic to whitelist, but only for the component that is dynamically derived from the spec (i.e I don't format the whole target, just the piece of it that comes from the spec).

I also added a test for this.

haampie commented on lib/spack/spack/cmd/env.py:
```diff
@@ -676,21 +677,42 @@ def build_cache_flag(self, depth):
+    Many characters in a typical Spec formatted string trigger special
+    behavior in `make`, including '=', and '%'.
+
+    Note that '-' is an acceptable character but may cause confusion when
+    invoking make from the shell; it is allowed here though.
```
```suggestion
```

I don't think it's worth commenting this, it tab completes fine

haampie commented on lib/spack/spack/cmd/env.py:
```diff
@@ -676,21 +677,42 @@ def build_cache_flag(self, depth):
+    Note that '-' is an acceptable character but may cause confusion when
+    invoking make from the shell; it is allowed here though.
+    """
+    if not isinstance(spec, spack.spec.Spec):
+        raise ValueError("Internal error: got non-spec object: {0}".format(spec))
```
Can you use a type annotation in the function argument and remove this? It's private api anyways

haampie commented on lib/spack/spack/cmd/env.py:
```diff
@@ -676,21 +677,42 @@ def build_cache_flag(self, depth):
+    if not isinstance(spec, spack.spec.Spec):
+        raise ValueError("Internal error: got non-spec object: {0}".format(spec))
+    tgt = spec.format("{name}-{version}-{hash}")
+    return re.sub(r"[^A-Za-z0-9_.-]", "_", tgt)
```
move to constructor and compile regex once?

scheibelp replied:
There is no unified constructor so I created a singleton object for this. I think without the singleton I'd need to modify the `MakeTargetVisitor.__init__` to add another attribute (which I'd be fine with).

In general, Python will cache patterns passed to `re.compile` so these changes feel excessive to me, but I don't think it overcomplicates things to reorganize this way.

haampie commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -3119,6 +3120,46 @@ def test_environment_depfile_makefile(depfile_flags, expected_installs, tmpdir,
+
+    monkeypatch.setattr(spack.spec.Spec, "format", _contrived_spec_format)
+
+    # Disable jobserver so we can do a dry run.
```
You're not using dry run 🤔 tbh, dry run is the best way to validate the makefile, and this is an integration test already, so it might even be better than manually validating the generated output

scheibelp replied:
Dry run could be useful to do in addition to this, but I strongly feel that it is not enough: the problem I ran into involved a makefile which you could use (no syntax errors, and running `make` proceeds), but was entirely wrong (e.g. would perform installations of packages in the wrong order); I don't think a dry run could catch this. It could be useful for detecting other problems with generating the makefile though.

haampie replied:
~Can you show an example? I don't entirely see how it would not result in make error~ 🤔 

I see now (commented in the wrong thread before, sorry)

```Makefile
abc=def: prereq # ordinary definition of `abc`
target: abc=def # target-specific variable value for variable `abc` in target `target`
```

Notice that a dry-run is more than "it runs", since it will print what it would do in order, which is better than parsing a Makefile. See other tests

scheibelp replied:
Is https://github.com/spack/spack/pull/36642/files#r1160315876 an answer to this? It is to me. The use case I worked from is enormous so I can't post the full file here but:

* It used a `hash=number` version for a build dependency
* `make` "skipped" installation of that and proceeded directly to installing packages that depended on this build dependency

That's why I think a dry run would not expose the error: you have to actually execute the task.

scheibelp replied:
I would need to check but I think the makefile generated by this test should successfully execute a dry run without the changes outside of the unit-tests (I didn't check that as part of this PR because (a) the special characters were the source of the problem in the larger example and (b) I removed those as part of the PR).

scheibelp replied:
To confirm: this test does in fact generate an example. If you `git checkout d2fd6807 -- lib/spack/spack/cmd/env.py ` (i.e. revert the core changes to before this PR) and add

```
    # Do make dry run.
    make = Executable("make")
    out = make("-n", "-f", makefile, output=str)
```

after the generation of the makefile in `test_environment_depfile_spec_format_special_chars` then it:

* "succeeds" (no error code)
* Doesn't ever attempt to install dtlink1

It does however do

`touch prefix/install/dtlink1-githash=version-spackhash`

which means that to confirm that the dtlink1 installation was missing, I would have to parse the actions (lines) with knowledge of the target format. This seems to me to be nearly as invasive as examining the makefile contents directly, and furthermore ignores other notions of correctness like proper installation order. The introduction of unexpected special characters could alter the makefile's behavior in many unexpected ways, and instead of verifying the makefile behavior I think it is more-appropriate to verify that no unexpected special characters are introduced.

haampie replied:
> instead of verifying the makefile behavior I think it is more-appropriate to verify that no unexpected special characters are introduced.

Sure, but imo an actual unit test would be best then: concrete specs in, adjacency list out; verify nodes, their names, and edges. That requires a bit of refactoring.

> This seems to me to be nearly as invasive

It's not. The minimal change is `2 insertions(+), 1 deletion(-)`, and could've been a oneliner if spack wasn't so insistent on needing a `git` prop in the mock package:

```diff
diff --git a/lib/spack/spack/test/cmd/env.py b/lib/spack/spack/test/cmd/env.py
index 8fb16e8322..662b87d74c 100644
--- a/lib/spack/spack/test/cmd/env.py
+++ b/lib/spack/spack/test/cmd/env.py
@@ -3091,7 +3091,7 @@ def test_environment_depfile_makefile(depfile_flags, expected_installs, tmpdir,
     make = Executable("make")
     makefile = str(tmpdir.join("Makefile"))
     with ev.read("test"):
-        add("dttop")
+        add("dttop ^dtlink1@aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa=1.0")
         concretize()
 
     # Disable jobserver so we can do a dry run.
diff --git a/var/spack/repos/builtin.mock/packages/dtlink1/package.py b/var/spack/repos/builtin.mock/packages/dtlink1/package.py
index 836ca59e69..a8c1d100b5 100644
--- a/var/spack/repos/builtin.mock/packages/dtlink1/package.py
+++ b/var/spack/repos/builtin.mock/packages/dtlink1/package.py
@@ -11,6 +11,7 @@ class Dtlink1(Package):
 
     homepage = "http://www.example.com"
     url = "http://www.example.com/dtlink1-1.0.tar.gz"
+    git = "http://www.example.com/dtlink1.git"
 
     version("1.0", "0123456789abcdef0123456789abcdef")
```

With a handful extra lines the topo order could also be tested. For example `set(children).issubset(specs_that_make_would_install[:i]))` for 0 <= i < `len(specs_that_make_would_install)`, and children is something with `traverse`

scheibelp replied:
> The minimal change is 2 insertions(+), 1 deletion(-), and could've been a oneliner if spack wasn't so insistent on needing a git prop in the mock package:

I want to be clear that I strongly recommend against doing anything like what this diff suggests: ultimately it does not matter what properties inside of specs give rise to special characters, only that they could be generated as part of using Spec functionality to generate a makefile target.

There is a better way to do what I did (and I'm happy to try to make that happen), but it is not this particular diff.

>> instead of verifying the makefile behavior I think it is more-appropriate to verify that no unexpected special characters are introduced.

> Sure, but imo an actual unit test would be best then: concrete specs in, adjacency list out; verify nodes, their names, and edges. That requires a bit of refactoring.

My initial interpretation of what you said is to examine something like `MakeTargetVisitor.adjacency_list`, but really any test that examines entities before the final makefile will be fragile: you have to know which properties to check; if the makefile generation introduces an extra reference to the spec anywhere then that would be another potential source of unwanted special characters.

I could test `_fmt_spec_make_target` directly (I think this would be closer to what you consider a unit test vs. an integration test), although if that isn't called in some spot where it should be, the makefile will again end up with special characters.

I could refactor the `env_depfile` function so that it only ever interacts with a new class like `MakefileSpec`, which could be mocked by the tests to ensure that users never call `str` on it, and only get special characters if they call an explicit function like `MakefileSpec.unsafe_format()`. Overall something like.

```
def env_depfile():
    roots = [s for _, s in env.concretized_specs()]
    all_specs = list(traverse(roots)) # No visitor here: need to convert type
    roots = [MakefileSpec(x) for x in roots]
    all_specs = [MakefileSpec(x) for x in all_specs]
    # At this point we have more control over how the makefile could pull details from the spec
    generate_makefile(roots, all_specs)
```

does that seem better?

haampie replied:
If you have an integration test for git versions, then this

> if the makefile generation introduces an extra reference to the spec anywhere then that would be another potential source of unwanted special characters.

is not a concern in the unit test.

If I would write it I would just make a small class (the model) that has properties that are passed to `template.render` (the view) and then unit test the model properties without dealing with the view at all. Also, no SpackCommand (the controller) would be necessary in that test.

haampie replied:
At the same time, I also don't care too much at this point 😆 as long as the bug fix gets in

scheibelp replied:
> If I would write it I would just make a small class (the model) that has properties that are passed to template.render (the view)

I think you're talking about taking the dictionary passed to `template.render` and turning it into an object:

```
    rendered = template.render(
        {
            "all_target": get_target("all"),
            "env_target": get_target("env"),
```

Validating such an object would be complex: the targets in general can have several different special characters, and while I don't think `=` is among them (which makes it easy to look for that specifically as a problem), there are other special characters that can be generated as part of `Spec.format` which *are* used in targets (e.g. `%`). Therefore, this object would need to differentiate between the spec and non-spec portions of targets, i.e. have functions like:

```
class Depfile(object):
    def add_spec_target(self, preamble: str, spec: Spec):
        # the preamble can contain "%" or anything we want, and isn't validated
        ...

    def add_target(self, target: str):
        # Hopefully no one called str() on a Spec to get the target
        ....
```

I think my suggestion from https://github.com/spack/spack/pull/36642#discussion_r1164510858 would simplify the matter:

> I could refactor the env_depfile function so that it only ever interacts with a new class like MakefileSpec (rather than Spec), 

The `MakefileSpecs` (which are not full specs and so can be mocked without monkeypatching) can be fed into the functions more or less the same as before to generate the template input dictionary; I could split the logic to then stop before rendering the template (to me, creating the template does not make this an integration test, although from discussion here I gather you think it does).

haampie replied:
> Validating such an object would be complex

I'm not sure I follow 🤔 I've implemented

> If I would write it I would just make a small class (the model) that has properties that are passed to template.render (the view)

in #36911, and it's not complex to do `data = MakefileModel(...).to_dict()` and test correctness of `data`. You could mock MakefileModel's `Spec` -> `str` function to produce something distinct and verify that it pops up in `data` for entries that are targets, etc, so you know the format function is called in the right places.

haampie commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -3119,6 +3120,46 @@ def test_environment_depfile_makefile(depfile_flags, expected_installs, tmpdir,
+        env(
+            "depfile",
+            "-o",
+            makefile,
```
You can save a few lines by dropping `-o makefile` and use what would be stdout (if not dry run)

haampie commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -3119,6 +3120,46 @@ def test_environment_depfile_makefile(depfile_flags, expected_installs, tmpdir,
+            "--use-buildcache=never",
+        )
+
+    with open(makefile, "r") as f:
+        makefile_lines = f.readlines()
+
+    makefile_contents = "\n".join(makefile_lines)
```
```suggestion
```

see above

haampie commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -3119,6 +3120,46 @@ def test_environment_depfile_makefile(depfile_flags, expected_installs, tmpdir,
+        add("dttop")
+        concretize()
+
+    orig_format = spack.spec.Spec.format
+
+    def _contrived_spec_format(obj, format_str):
+        if obj.name == "dtlink1":
+            return "dtlink1-githash=version-spackhash"
+        else:
+            return orig_format(obj, format_str)
+
+    monkeypatch.setattr(spack.spec.Spec, "format", _contrived_spec_format)
```
Since you already have mock_packages, can't you actually use a normal git version with user specified spack version to the same effect? That would be less contrived

scheibelp replied:
I wanted to avoid patching the packages to achieve this condition: IMO the source of all errors like the one this PR aims to correct is specially-interpreted characters via `Spec.format`; therefore I argue that this is the least-contrived way to test the outcome (changing things about the versions etc. would silently render this test useless if `spec.format` happened to behave differently in some scenarios vs. others).

haampie replied:
Hm, it's a rule of thumb not to mock in integration tests, cause assumptions on the implementation make the test redundant on refactor. If you want an integration test for git versions, then just use a git version. If you want to test normalization of targets, then write a unit test. In both cases you don't need to use monkeypatch

haampie replied:
Ah... I see what the issue is, or at least I think I do :p we're generating target/prereqs without rules like

```
<pkg a>-<version a>-<hash a>: <pkg b>-<version b>-<hash b>
```

so presumably if version contains an `=` it parses as a variable declaration

scheibelp replied:
>  If you want to test normalization of targets

I don't want to normalize the entire target: the depfile uses special characters in targets for their intended effect. The issue is when those targets are partially defined using `Spec.format`, which introduces *unexpected* special characters.

> If you want an integration test for git versions, then just use a git version.

I don't think git versions are the (real) problem. For a system that generates makefiles of spack installs and needs to create targets based on the Specs, each such target is composed of:

* Static components relevant to makefile processing
* Details extracted from the Spec itself

The latter is the only place that can surprise us. The current logic has no formalized abstraction for the latter and directly uses `.format`. I considered it appropriate to just alter `Spec.format`, but it would be more-robust to mock the Spec object entirely.

> Hm, it's a rule of thumb not to mock in integration tests, cause assumptions on the implementation make the test redundant on refactor.

given the above, while I'd say it is strictly true that this mocking makes an assumption in the implementation, setting a git version is even more a violation of this rule (conditional changes to `str()` of version could render this test useless: why not target the point where the depfile logic directly interfaces with the Spec API?).

haampie replied:
I'm sorry for being so insistent, but I'd rather see `monkeypatch` dropped. The other point is that `Spec.format` is so commonly called, mocking it will likely break invariants elsewhere in the future.

Please just add an integration/regression test with depfile and git versions which is the problem you're solving in the PR (you can just adapt the test above the one you've added), and if you don't like integration tests, refactor so that you can unit test the adjacency list

scheibelp committed:
update comment

scheibelp committed:
enforce types with annotation; another docstring update

scheibelp committed:
pre-compile regex pattern

scheibelp committed:
encapsulate regex compilation into singleton object: avoid compiling on module import while still only compiling the pattern once

scheibelp committed:
style fix and type annotation fix
